canonical/istio-k8s-operator
# Add `charm-tracing` and `workload-tracing` support

This PR adds support for charm tracing as well as istio-managed workload (L7) traces using OTLP. It includes adding the tracing endpoints as well as constructing the tracing config with an `otlp_grpc` endpoint sent by a tracing backend (i.e Tempo). 

## Context
- For traces to be sent from Istio to an on-mesh Tempo, https://github.com/canonical/istio-k8s-operator/issues/30 must be fixed 
- In Istio, the [waypoint](https://github.com/canonical/istio-beacon-k8s-operator) is the component responsible for generating traces/spans and exporting them to a tracing backend. This requires that the Istio Beacon charm (https://github.com/canonical/istio-beacon-k8s-operator) is to be deployed on a model with `model-on-mesh=true` to generate traces from traffic in that model.
- `Istio-core`  charm needs to be deployed using `auto-allow-waypoint-policy=false`

## Testing Considerations
1. `juju add-model istio-system`
2. `juju deploy ./istio-k8s_ubuntu-22.04-amd64.charm istio --config auto-allow-waypoint-policy=false --trust`
3. `juju add-model mesh`
4. Deploy Tempo HA
5. `juju deploy istio-beacon-k8s --channel latest/edge --config model-on-mesh=true --trust` 
6. `juju deploy grafana-k8s grafana --trust --channel=edge`
7. `jhack imatrix fill`
8. `juju offer tempo:tracing`
9. `juju switch istio-system`
10. `juju consume admin/mesh.tempo`
11. `juju integrate istio-k8s:charm-tracing tempo`
12. `juju integrate istio-k8s:workload-tracing tempo`
13. if https://github.com/canonical/istio-k8s-operator/issues/30  is fixed, skip to step 14
`microk8s.kubectl apply -f serviceentry.yaml`
serviceentry.yaml:
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: tempo
spec:
  hosts:
  - tempo-0.tempo-endpoints.mesh.svc.cluster.local
  ports:
  - number: 4317
    name: grpc-otel
    protocol: GRPC
  resolution: DNS
```
14. Open Grafana -> sources -> tempo
15. you should see traces under `istio-beacon-k8s-mesh-waypoint.mesh` service name
